src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,750 @@
+                    return;
+                }
+
+                _lastProcessedStreamId = lastProcessedStreamId;
+
+                foreach (KeyValuePair<long, Http3RequestStream> request in _activeRequests)
+                {
+                    if (request.Key > lastProcessedStreamId)
+                    {
+                        streamsToGoAway.Add(request.Value);
+                    }
+                }
+
+                CancelWaiters();
+                CheckForShutdown();
+            }
+
+            // GOAWAY each stream outside of the lock, so they can acquire the lock to remove themselves from _activeRequests.
+            foreach (Http3RequestStream stream in streamsToGoAway)
+            {
+                stream.GoAway();
+            }
+        }
+
+        public void RemoveStream(long streamId)
+        {
+            lock (SyncObj)
+            {
+                bool removed = _activeRequests.Remove(streamId);
+                Debug.Assert(removed == true);
+
+                if (ShuttingDown)
+                {
+                    CheckForShutdown();
+                }
+            }
+        }
+
+        public override void Trace(string message, [CallerMemberName] string memberName = null) =>
+            Trace(0, message, memberName);
+
+        internal void Trace(long streamId, string message, [CallerMemberName] string memberName = null) =>
+            NetEventSource.Log.HandlerMessage(
+                _pool?.GetHashCode() ?? 0,    // pool ID
+                GetHashCode(),                // connection ID
+                (int)streamId,                // stream ID
+                memberName,                   // method name
+                message);                     // message
+
+        private async ValueTask SendSettingsAsync()
+        {
+            try
+            {
+                _clientControl = _connection.OpenUnidirectionalStream();
+                await _clientControl.WriteAsync(_pool.Settings.Http3SettingsFrame, CancellationToken.None).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Abort(ex);
+            }
+        }
+
+        public static byte[] BuildSettingsFrame(HttpConnectionSettings settings)
+        {
+            Span<byte> buffer = stackalloc byte[4 + VariableLengthIntegerHelper.MaximumEncodedLength];
+
+            int integerLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(4), settings._maxResponseHeadersLength * 1024L);
+            int payloadLength = 1 + integerLength; // includes the setting ID and the integer value.
+            Debug.Assert(payloadLength <= VariableLengthIntegerHelper.OneByteLimit);
+
+            buffer[0] = (byte)Http3StreamType.Control;
+            buffer[1] = (byte)Http3FrameType.Settings;
+            buffer[2] = (byte)payloadLength;
+            buffer[3] = (byte)Http3SettingType.MaxHeaderListSize;
+
+            return buffer.Slice(4 + integerLength).ToArray();
```

Comment:
Allocating this byte[] is ok because it's rare?

Reply:
Yes. Once per `SocketsHttpHandler`.